### ISSUE TITLE
"Low-level Error: Symbol extends beyond the end of the address space" in caller of a vararg function

### ISSUE THREAD

**Issue by CreepNT** [NationalSecurityAgency/ghidra#4148]
**Describe the bug**
During decompilation of a caller function from an ARMv7 binary, decompilation window will show `Low-level Error: Symbol $$undef0000000X extends beyond the end of the address space` instead of decompiling properly. This *might* be due to bad stack argument detection, as the issue goes away if *varargs* is removed from called function, but reappears if overriding signature with "too much" arguments.

**To Reproduce**
1. Import `sysmem.elf` using the standard ELF loader (`ARM:LE:32:v7`, `default` compiler ID)
2. Go to address `0x81021340`, and disassemble function located there (pressing F12 - the function is Thumb mode) - decompilation fails.
3. Find call to `FUN_8102ed24`, and change its signature to `void FUN_8102ed24(int level, char* fmt)` - function now decompiles properly
4. Add *Varargs* to `FUN_8102ed24`'s signature - decompilation fails again

**Expected behavior**
Function decompiles properly.

**Screenshots**
With incomplete overridden signature: decompiles properly, but arguments `index`, `addr` and `size` are missing.
![Incomplete overridden signature](https://user-images.githubusercontent.com/32939301/162737328-6fbefed8-bc5f-4516-8523-14311050099e.png)

Adding just a single argument to the override results in decompilation error.
![Overridden signature failure](https://user-images.githubusercontent.com/32939301/162740520-a2bf2e96-7d6f-4707-832e-ed6d11e74ef1.png)

Proper override (`void SceDebugForDriver_1A3F2AA4(int level, char * fmt, int, char *, void *, int, int, int)`) results in the same error.
![Proper overriden signature failure](https://user-images.githubusercontent.com/32939301/162740799-5b97adb3-c33e-4209-af8c-03fac8926f2d.png)

**Attachments**
Archive with binary that has faulty function: [sysmem.zip](https://github.com/NationalSecurityAgency/ghidra/files/8464117/sysmem.zip)

**Environment:**
 - OS: Windows 10 20H2 19042.1586
 - Java Version: 11.0.5
 - Ghidra Version: 10.1.2
 - Ghidra Origin: official, from this repository's `Releases` tab

**Comment by CreepNT** [comment]
An update on this bug: I have encountered the same problem in a different revision of the same binary, in the same function.
I looked into the issue more closely and might have found the underlying problem causing this bug.

This code sequence is executed when an invalid `pPage` pointer is given - more specifically, if `pPage == 0xFFFFFFFF`.
But in this sequence, the program attempts to print information obtained via the pointer.

A concise summary of the assembly is:
```armasm
    @r11 holds pPage
    cmp r11, #0xFFFFFFFF
    beq bad_pPage
    @ ... 

bad_pPage:
    @ ...
    ldr r7, [r11, #0] @Attempts to read a u32 at 0xFFFFFFFF (i.e. pPage->unk0)
    @ ...

    @Prepare arguments for printf()
    str r7, [sp, #0xC]
    str r4, [sp, #0x8]
    str r11, [sp, #0x0]
    str r6, [sp, #0x4]
    bl sceKernelPrintfLevel
```

Once translated to C, the sequence (including the omitted parts) would give code akin to
```c
if (pPage == 0xFFFFFFFF) {
    //a first printf() without arguments, and irrelevant
    sceKernelPrintfLevel(0x1, "%d:phymem.c:%s:Before:pPage %p, index=0x%08x, addr=0x%08x, size=0x%08x\n",
       sceKernelCpuId(), "PhyMemPartFreePhyPageBuddyRaw", pPage /* optimized by the compiler to its value, 0xFFFFFFFF */,
       pPage->unk0 & 0xFFFFF, (pPage->unk0 & 0xFFFFF) << 12, 1 << ((pPage->unk0 >> 20) & 0x1F) );
}
```

TL;DR the bug seems related to memory accesses done via an invalid pointer with value 0xFFFFFFFF.

I confirmed the bug to still be present in Ghidra 10.2.3 (with the program first imported in 10.1, if it may be relevant), although the missing symbol name has changed from `$$undef00000006` instead of `$$undef00000004`.

**Comment by 46cv8** [comment]
Is there anyway to get additional debugging information regarding what line of assembler/byte code was being processed at the time the error occurred, without running Ghidra under a java debugger?

**Comment by CreepNT** [comment]
Seeing the following entry in the 12.1.3 changelog:

> Decompiler. Added Decompiler support for memory accesses that wrap from high to low addresses. (GP-6936)

I was hoping the issue would be solved, but alas the decompilation still fails, although with a different error now (which at least is understandable):

```
  Low-level Error: Trying to construct memory range beyond end of address space: ram
```

**Comment by CBongo** [comment]
@CreepNT GP-6936 cannot fix this as shipped, which explains what you saw on 12.1.3. `allows_wrapped_range` is set in only two places, both `SpacebaseSpace` constructors (`translate.cc:64`  and `:80`), with no decode path, so a `ram` space can never carry it and `constructWrappingAddress`'s throw is unconditional for exactly your case.

I reproduced your error on `sysmem.elf` at 0x81021340 on stock 12.1.3, and granting the flag to non-truncated processor spaces makes it decompile cleanly, with `pPage->unk0` recovered as `CONCAT31(uRam00000000,uRamffffffff)`, matching the expectation in your 2023 comment. Filed as PR #9656  with the patch, since the same root cause also regresses every 6502 call site.
